base_layer/core/src/chain_storage/blockchain_database.rs

```diff
@@ -129,25 +129,34 @@ impl<B: BlockchainBackend> Clone for Validators<B> {
 /// operation within it must succeed, or they all fail. Failure to support this contract could lead to
 /// synchronisation issues in your database backend.
 ///
-/// Data is passed to and from the backend via the [DbKey], [DbValue], and [DbValueKey] enums. This strategy allows
-/// us to keep the reading and writing API extremely simple. Extending the types of data that the back ends can handle
-/// will entail adding to those enums, and the back ends, while this trait can remain unchanged.
+/// This trait is written so that every it is assumed that every function call should be atomic, e.i. if some part of
+/// the instruction fails, the entire call should be reversed
 pub trait BlockchainBackend: Send + Sync {
-    /// Commit the transaction given to the backend. If there is an error, the transaction must be rolled back, and
-    /// the error condition returned. On success, every operation in the transaction will have been committed, and
-    /// the function will return `Ok(())`.
-    fn write(&mut self, tx: DbTransaction) -> Result<(), ChainStorageError>;
+    /// Adds a block to the orphan database. This function assumes that the stateless validation has passed on the
+    /// block.
+    fn add_orphan_block(&mut self, block: Block) -> Result<(), ChainStorageError>;
+    /// This function will move a block from the orphan pool to the main chain. It assumes that the block has passed
+    /// full stateful validation
+    fn accept_block(&mut self, block_hash: HashOutput) -> Result<(), ChainStorageError>;
+    // rewinds the database to the specified height. It will move every block that was rewound to the orphan pool
+    fn rewind_to_height(&mut self, height: u64) -> Result<(), ChainStorageError>;
+    /// This is used when synchronising. Adds in the list of headers provided to the main chain
+    fn add_block_headers(&mut self, headers: Vec<BlockHeader>) -> Result<(), ChainStorageError>;
+    /// This is used when synchronising. Adds in the list of kernels provided to the main chain
+    fn add_kernels(&mut self, kernels: Vec<TransactionKernel>) -> Result<(), ChainStorageError>;
+    /// This is used when synchronising. Adds in the list of utxos provided to the main chain
+    fn add_utxos(&mut self, utxos: Vec<TransactionOutput>) -> Result<(), ChainStorageError>;
+    /// This is used when synchronising. Adds in the list of mmr leafs provided to the main chain
+    fn add_mmr(&mut self, tree: MmrTree, hashes: Vec<HashOutput>) -> Result<(), ChainStorageError>;
     /// Fetch a value from the back end corresponding to the given key. If the value is not found, `get` must return
     /// `Ok(None)`. It should only error if there is an access or integrity issue with the underlying back end.
     fn fetch(&self, key: &DbKey) -> Result<Option<DbValue>, ChainStorageError>;
```

Comment:
In postgres, you'd probably have to switch between 5 different tables to implement this fetch depending on `DbKey`. Perhaps we should split into `fetch_block(BlockKey::Hash(xxxx))` or `fetch_block(BlockKey::Height(123))` - `fetch_metadata(MetadataKey::ChainHeight)` etc.
```rust
pub enum DbKey {
    Metadata(MetadataKey),
    BlockHeader(u64),
    BlockHash(BlockHash),
    UnspentOutput(HashOutput),
    SpentOutput(HashOutput),
    TransactionKernel(HashOutput),
    OrphanBlock(HashOutput),
}
```

Reply:
Same for contains, though we should only add the functions that are used.

Reply:
Spoken to @SWvheerden the implementation is pretty clean without splitting these up :+1: